diffkemp/simpll/DifferentialFunctionComparator.cpp

```diff
@@ -1959,3 +1985,21 @@ bool DifferentialFunctionComparator::collectBinaryOperands(
     }
     return false;
 }
+
+/// Recursively check if users of the instruction with an inverse condition
+/// are only branch or not instructions.
+/// This is done to prevent false negatives.
+/// Return true if the users are acceptable for inverse conditions pattern,
+/// false otherwise.
+bool DifferentialFunctionComparator::checkInverseCondUsers(
+        const Instruction *inst) const {
+    for (auto user : inst->users()) {
+        auto userInstruction = dyn_cast<Instruction>(user);
+        if (isa<BranchInst>(userInstruction) || isLogicalNot(userInstruction)) {
+            checkInverseCondUsers(userInstruction);
+        } else {
+            return false;
+        }
```

Comment:
This doesn't seem correct, you're not using the result of the recursive call. Also no need for the recursive call in case of `BranchInst`.
Perhaps should be like this?
```suggestion
        if (isLogicalNot(userInstruction) && !checkInverseCondUsers(userInstruction)) {
            return false;
        } else if (!isa<BranchInst>(userInstruction)) {
            return false;
        }
```
We could possibly join the conditions together, but it may be less readable.

Reply:
Thanks, I somehow did not realise or overlooked that...
I used your suggestion but slightly modified it (there was a problem when it was `not` instruction).

Reply:
Good point, I see it now.